src/blockrelay/compactblock.cpp

```diff
@@ -0,0 +1,1198 @@
+            nIndex++;
+        }
+        CompactReRequest compactReRequest;
+        compactReRequest.blockhash = header.GetHash();
+        compactReRequest.indexes = vIndexesToRequest;
+        pfrom->PushMessage(NetMsgType::GETBLOCKTXN, compactReRequest);
+
+        // Update run-time statistics of compact block bandwidth savings
+        compactdata.UpdateInBoundReRequestedTx(pfrom->compactBlockWaitingForTxns);
+        return true;
+    }
+
+    // If there are still any missing transactions then we must clear out the compactblock data
+    // and re-request a full block (This should never happen because we just checked the various pools).
+    if (missingCount > 0)
+    {
+        // Since we can't process this compactblock then clear out the data from memory
+        compactdata.ClearCompactBlockData(pfrom, header.GetHash());
+
+        thinrelay.RequestBlock(pfrom, header.GetHash());
+        return error("Still missing transactions for compactblock: re-requesting a full block");
+    }
+
+    // We now have all the transactions now that are in this block
+    pfrom->compactBlockWaitingForTxns = -1;
+    int blockSize = pfrom->compactBlock.GetBlockSize();
+    LOG(CMPCT, "Reassembled compactblock for %s (%d bytes). Message was %d bytes, compression ratio %3.2f, peer=%s\n",
+        pfrom->compactBlock.GetHash().ToString(), blockSize, pfrom->nSizeCompactBlock,
+        ((float)blockSize) / ((float)pfrom->nSizeCompactBlock), pfrom->GetLogName());
+
+    // Update run-time statistics of compact block bandwidth savings
+    compactdata.UpdateInBound(pfrom->nSizeCompactBlock, blockSize);
+    LOG(CMPCT, "compact block stats: %s\n", compactdata.ToString().c_str());
+
+    // Process the full block
+    PV->HandleBlockMessage(pfrom, NetMsgType::CMPCTBLOCK, MakeBlockRef(pfrom->compactBlock), GetInv());
+
+    return true;
+}
+
+bool CompactReRequest::HandleMessage(CDataStream &vRecv, CNode *pfrom)
+{
+    CompactReRequest compactReRequest;
+    vRecv >> compactReRequest;
+
+    // Message consistency checking
+    if (compactReRequest.indexes.empty() || compactReRequest.blockhash.IsNull())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("incorrectly constructed getblocktxn received.  Banning peer=%s", pfrom->GetLogName());
+    }
+
+    // We use MSG_TX here even though we refer to blockhash because we need to track
+    // how many xblocktx requests we make in case of DOS
+    CInv inv(MSG_TX, compactReRequest.blockhash);
+    LOG(CMPCT, "received getblocktxn for %s peer=%s\n", inv.hash.ToString(), pfrom->GetLogName());
+
+    std::vector<CTransaction> vTx;
+    CBlockIndex *hdr = LookupBlockIndex(inv.hash);
+    if (!hdr)
+    {
+        dosMan.Misbehaving(pfrom, 20);
+        return error("Requested block is not available");
+    }
+    else
+    {
+        CBlock block;
+        const Consensus::Params &consensusParams = Params().GetConsensus();
+        if (!ReadBlockFromDisk(block, hdr, consensusParams))
```

Comment:
before reading the block from disk, we should make sure its near the tip and perhaps limit the number of rerequests per 10 minutes per node.